### DISCUSSION TITLE
[HSRB] Hand Camera Connection Error - Image Not Displayed

### DISCUSSION THREAD

**Discussion by MrKeith99** [HumanSupportRobot/QandA#106] [Q&A]
**Detected Problem:**

I'm encountering an issue with the hand camera connection in my HSR robot. The diagnostic tool reports a "Connection error," and while Cheese in Ubuntu recognizes the camera, it doesn't display an image.

This is the console result when trying to change the camera to the hand camera in Cheese:
```
(cheese:6788): cheese-WARNING **: 23:33:16.360: Failed to allocate required memory.: gstv4l2src.c(660): gst_v4l2src_decide_allocation (): /GstCameraBin:camerabin/GstWrapperCameraBinSrc:camera_source/GstBin:bin21/GstV4l2Src:v4l2src4:
Buffer pool activation failed
```

> [!NOTE]
> In the 24.10 version (other versions might be the same), when launching Cheese, camera selection cannot be conducted as the property function is not shown. We had to reinstall Cheese as described in [AskUbuntu](https://askubuntu.com/questions/1133058/how-to-access-the-missing-settings-preferences-menu-in-cheese-in-non-gnome-des)

**Troubleshooting Steps Taken:**

* **HSRB information (clean setup):**
  * RELEASE_VERSION=24.10.01 # same problem in 23.10.01
  * ROBOT_VERSION=HSRB-PHASE5
  * LOCALE=ja_JP.UTF-8

* **Verified device detection:**
  * `lsusb` confirms the camera is detected by the system.
  * Cheese recognizes the camera but doesn't show an image.
* **Checked physical connection:**
  * Inspected the USB connection within the hand compartment.
  * Tried a different USB cable on another computer, and the camera functions properly.
* **Tested on robot's USB3.0 ports:**
  * The camera displays an image in Cheese when connected directly to the robot's rear USB port.
  * However, the ros node fails to launch when the emergency button is unpressed.
* **Bandwidth research:**
  * Investigated the possibility of USB 2.0 bandwidth limitations, as shown in [Nvidia Developers Forum](https://forums.developer.nvidia.com/t/use-gstreamer-to-open-two-usb-cameras-failed-to-allocate-required-memory/122843)
  * Uncertain why this wouldn't affect other HSR users.

**Current Considerations:**

* Replacing the USB cable might be necessary, but opening the robot cover could be needed and apparently this is not allowed because of the contract.

**Request for Assistance:**

* Due to our team's participation in the [HSRT-X](https://www.hsr.io/archives/428) project, we'd appreciate any guidance on resolving this issue without having to send to repair the robot. 
* If self-repair isn't recommended, please advise on the appropriate course of action.

**Comment by MrKeith99** [comment]
Btw, I have also checked the [Hand Camera Troubleshooting](https://docs.hsr.io/hsrb_maintenance_manual_en/src/check_hand_camera.html?highlight=hand)

**Comment by yoshimi-iyoda** [comment]
Thank you for the detailed information.
Regarding opening the robot's cover, it used to be uniformly prohibited by the contract, but recently, I believe the restrictions have been eased depending on the contract partner, so please confirm this.

As for countermeasures, as mentioned below, you might consider connecting a relay USB hub or replacing the cable with an external one.
https://docs.hsr.io/hsr_faq_en/hsr_common_faq.html#id9

**Reply by MrKeith99** [reply]
Thank you for your replay.

In the end, we decided to replace the cable with an external one.

Let me share how we did it, so that other people who are facing with the same problem, may try this method.

**Successfully Bypassed Internal Cable with External Solution**

This guide details the process of replacing the internal USB cable of the HSR hand camera with an external one. This approach offers a solution for users facing image transfer issues potentially caused by USB 2.0 bandwidth limitations.

**Troubleshooting Steps:**

> [!Important]
> Be sure to have the HSR emergency stop button pressed when following these changes.

1. **Identifying the Faulty Cable:**
   - We followed the instructions in the [Hand Camera Troubleshooting Guide](https://docs.hsr.io/hsrb_maintenance_manual_en/src/check_hand_camera.html).
   - This process helped us identify cable `1011` as the culprit for failing to transfer images to the CPU.
   - We believe the bandwidth limitation of USB 2.0 might be the cause, as discussed in the [Nvidia Developers Forum](https://forums.developer.nvidia.com/t/use-gstreamer-to-open-two-usb-cameras-failed-to-allocate-required-memory/122843).

> [!Note]
> In some cases, reinstalling the Ubuntu `Cheese` program might be necessary to verify camera functionality. Refer to the [AskUbuntu](https://askubuntu.com/questions/1133058/how-to-access-the-missing-settings-preferences-menu-in-cheese-in-non-gnome-des) guide for instructions.

2. **Opening the Hand and Replacing the Cable:**
   - We followed the [Hand Motor Board Replacement Procedure](https://docs.hsr.io/hsrb_maintenance_manual_en/src/replace_hand_motor_driver.html) up to `Step 3` to access the hand and replace the internal USB cable with an external one.

3. **Connecting and Testing the External Cable:**
   - Connect the new USB cable to a free USB port on the HSR access panel.

4. **Verifying Camera Functionality:**
   - Use the `Cheese` application to confirm the camera is working correctly.

5. **Fixing the Hand Cover:**
   - Once functionality is confirmed, carefully reattach the hand cover.

> [!Note]
> We used the small gap between cover and the actuator to pass the USB cable.

6. **Resolving ROS Camera Visualization Issue:**
   - While the `hand_camera` might now work in `Cheese`, ROS connection for camera visualization might still fail.
   - This is because the default rule files might conflict with the `head_center_camera`. We addressed this by modifying the rules as follows:
     a. Connect to the HSR via SSH as the `administrator` user.
     b. Navigate to the rules folder: `cd /etc/udev/rules.d`
     c. Use the `ls` command to confirm the presence of the `40-webcam.rules` file. 
     d. Create a backup of the file in case you need to revert the changes:  `sudo cp 40-webcam.rules 40-webcam.rules.backup`
     e. Edit the contents of `40-webcam.rules` using a text editor such as `nano` or `gedit` with `sudo` privileges. Make the following changes:

      ```diff
      - KERNEL=="video[0-9]*", KERNELS=="1-?*", SUBSYSTEMS=="usb", ATTRS{idVendor}=="1742", ENV{MINOR}=="0", SYMLINK+="hand_camera"
      - KERNEL=="video[0-9]*", KERNELS=="3-?*", SUBSYSTEMS=="usb", ATTRS{idVendor}=="1742", ENV{MINOR}=="2", SYMLINK+="head_center_camera"
      + KERNEL=="video[0-9]*", KERNELS=="1-?*", SUBSYSTEMS=="usb", ATTRS{idVendor}=="1742", ENV{MINOR}=="0", SYMLINK+="head_center_camera"
      + KERNEL=="video[0-9]*", KERNELS=="3-?*", SUBSYSTEMS=="usb", ATTRS{idVendor}=="1742", ENV{MINOR}=="2", SYMLINK+="hand_camera"
      ```
    f. Reload the updated rules with: `sudo udevadm control --reload-rules`, once changes on the `40-webcam.rules` file are saved.

7. **Disabling `head_center_camera` in ROS Configuration:**
   - Edit the `/etc/opt/tmc/robot/docker.hsrb.user` file and set the `USE_HEAD_CENTER_CAMERA` variable to `false`. This disables the `head_center_cable`, preventing interference with the PS eyefish camera connected via the external USB cable.

   ```
   # USE_HAND_CAMERA=false
   USE_HEAD_CENTER_CAMERA=false
   ```

> [!Note]
> Also check that `USE_HAND_CAMERA` is commented out or set to `true`.

8. Enabling the configuration with `sudo systemctl restart docker.hsrb.monitor.service`
9. Release the emergency stop button and start up the HSR to check `hand_camera` topic is being published properly.

We hope this instructions helps other HSR users in order to enable the visualization of the `hand_camera`
If you have any other question related to this topic. please let us know.
